voxblox/include/voxblox/io/layer_io_inl.h

```diff
@@ -0,0 +1,248 @@
+#ifndef VOXBLOX_CORE_IO_LAYER_IO_INL_H_
+#define VOXBLOX_CORE_IO_LAYER_IO_INL_H_
+
+#include <fstream>
+
+#include "./Block.pb.h"
+#include "./Layer.pb.h"
+
+#include "voxblox/utils/protobuf_utils.h"
+
+namespace voxblox {
+namespace io {
+
+template <typename VoxelType>
+bool LoadBlocksFromFile(
+    const std::string& file_path,
+    typename Layer<VoxelType>::BlockMergingStrategy strategy,
+    bool multiple_layer_support, Layer<VoxelType>* layer_ptr) {
+  CHECK_NOTNULL(layer_ptr);
+  CHECK(!file_path.empty());
+
+  // Open and check the file
+  std::fstream proto_file;
+  proto_file.open(file_path, std::fstream::in);
+  if (!proto_file.is_open()) {
+    LOG(ERROR) << "Could not open protobuf file to load layer: " << file_path;
+    return false;
+  }
+
+  // Byte offset result, used to keep track where we are in the file if
+  // necessary.
+  uint32_t tmp_byte_offset = 0;
+
+  bool layer_found = false;
+
+  do {
+    // Get number of messages
+    uint32_t num_protos;
+    if (!utils::readProtoMsgCountToStream(&proto_file, &num_protos,
+                                          &tmp_byte_offset)) {
+      LOG(ERROR) << "Could not read number of messages.";
+      return false;
+    }
+
+    if (num_protos == 0u) {
+      LOG(WARNING) << "Empty protobuf file!";
+      return false;
+    }
+
+    // Get header and check if it is compatible with existing layer.
+    LayerProto layer_proto;
+    if (!utils::readProtoMsgFromStream(&proto_file, &layer_proto,
+                                       &tmp_byte_offset)) {
+      LOG(ERROR) << "Could not read layer protobuf message.";
+      return false;
+    }
+
+    if (layer_ptr->isCompatible(layer_proto)) {
+      layer_found = true;
+    } else if (!multiple_layer_support) {
+      LOG(ERROR)
+          << "The layer information read from file is not compatible with "
+             "the current layer!";
+      return false;
+    } else {
+      // Figure out how much offset to jump forward by. This is the number of
+      // blocks * the block size... Actually maybe we just read them in? This
+      // is probably easiest. Then if there's corrupted blocks, we abort.
+      // We just don't add these to the layer.
```

Comment:
If you need that extra speed boost, why are you reading protobuf layers from file with multiple layer support? ;)

Reply:
Also this way you avoid the problem you were really concerned about: being 1 off on the size of the skipped region, etc. I'd rather trade a bit of speed on a reading-from-disk function rather than lose robustness by doing crazy math.